### PR TITLE
Fix XAddOptions maxlen handling and XPendingOptions validation

### LINKED ISSUE #2982: Negative maxlen for XAddOptions is ignored
This issue refers to a comment in [#2936](https://github.com/spring-projects/spring-data-redis/pull/2936#issuecomment-2220383705)

As we now can use XAddOptions in streamOperations,

```java
XAddOptions options = XAddOptions.maxlen(-1);
redisTemplate.opsForStream("myStream", Map.of("key", "value"), options);
```

This executes without throwing an error or warning the client. Instead, the result is that the maxlen option is not applied.

IMO, this is because the XAddOptions hasMaxlen() method gets called before every time XAddOptions is used, and it checks maxlen > 0, which results in ignoring negative values for maxlen.

- Reproducible code

```java
@ParameterizedRedisTest
void addMinusMaxlenDoesNotThrowsException() {
	K key = keyFactory.instance();
	HV value = hashValueFactory.instance();
	XAddOptions options = XAddOptions.maxlen(-1).approximateTrimming(false);
	assertDoesNotThrow(() -> redisTemplate.opsForStream().add(StreamRecords.objectBacked(value).withStreamKey(key), options));
	assertThat(redisTemplate.opsForStream().range(key, Range.unbounded())).hasSize(1);
}
```
image

### PR DESCRIPTION
**Pull request by jinkshower**
Closes #2982

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: src/main/java/org/springframework/data/redis/connection/stream/StreamReadOptions.java [hunk 1]
```diff
@@ -96,7 +96,7 @@ public StreamReadOptions block(Duration timeout) {
 	 */
 	public StreamReadOptions count(long count) {
 
-		Assert.isTrue(count > 0, "Count must be greater or equal to zero");
+		Assert.isTrue(count > 0, "Count must be greater than zero");
```
Comment (jinkshower) [inline src/main/java/org/springframework/data/redis/connection/stream/StreamReadOptions.java:99]: fix description to represent assertion

File: src/main/java/org/springframework/data/redis/connection/ReactiveStreamCommands.java [hunk 2]
```diff
@@ -337,7 +337,7 @@ public Long getMaxlen() {
 		 * @since 2.3
 		 */
 		public boolean hasMaxlen() {
-			return maxlen != null && maxlen > 0;
+			return maxlen != null;
```
Comment (jinkshower) [inline src/main/java/org/springframework/data/redis/connection/ReactiveStreamCommands.java:340]: For context, Lettuce and Jedis treat `maxlen` edge case differently. `XAddArgs` in Lettuce asserts `maxlen > 0` whereas Jedis doesn't check on `maxlen` so it follows default redis behaviour which asserts `maxlen >=0`.  My best choice was just let the user know. So I didn't put any assertion on `maxlen`.

```
java.lang.IllegalArgumentException: Maxlen must be greater 0

	at io.lettuce.core.internal.LettuceAssert.isTrue(LettuceAssert.java:193)
	at io.lettuce.core.XAddArgs.maxlen(XAddArgs.java:121)
	at org.springframework.data.redis.connection.lettuce.LettuceStreamCommands.xAdd(LettuceStreamCommands.java:84)
	at org.springframework.data.redis.connection.DefaultedRedisConnection.xAdd(DefaultedRedisConnection.java:489)
	at org.springframework.data.redis.connection.DefaultStringRedisConnection.xAdd(DefaultStringRedisConnection.java:2918)
	at org.springframework.data.redis.core.DefaultStreamOperations.lambda$add$2(DefaultStreamOperations.java:153)
```

File: src/test/java/org/springframework/data/redis/core/DefaultStreamOperationsIntegrationTests.java [hunk 3]
```diff
@@ -206,6 +206,19 @@ void addMaxLenShouldLimitSimpleMessagesSize() {
...
 
+	@ParameterizedRedisTest // GH-2982
+	void addNegativeMaxlenShouldThrowException() {
+
+		K key = keyFactory.instance();
+		HV value = hashValueFactory.instance();
+
+		XAddOptions options = XAddOptions.maxlen(-1).approximateTrimming(false);
+
+		assertThatThrownBy(() -> streamOps.add(StreamRecords.objectBacked(value).withStreamKey(key), options));
```
Comment (mp911de) [inline src/test/java/org/springframework/data/redis/core/DefaultStreamOperationsIntegrationTests.java:217]: Testing Redis/Driver behavior exceeds the scope of what we want to test. We will remove these tests during the merge.

**Review by mp911de — APPROVED** [review]

**Comment by mp911de** [comment]
Thank you for your contribution. That's merged, polished, and backported now.

**Comment by jinkshower** [comment]
Thank you for the rapid review! I've learned so much doing this and previous PR by comprehending and reading source codes!

**Pull request closed without merging**